File: app/src/main/java/com/an/github/ui/activity/GithubListActivity.java
package com.an.github.ui.activity;

import android.arch.lifecycle.ViewModelProviders;
import android.databinding.DataBindingUtil;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v4.app.ActivityOptionsCompat;
import android.support.v4.util.Pair;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.LinearLayoutManager;
import android.view.View;
import com.an.github.R;
import com.an.github.data.local.entity.GithubEntity;
import com.an.github.databinding.GithubListActivityBinding;
import com.an.github.factory.ViewModelFactory;
import com.an.github.ui.adapter.FilterListAdapter;
import com.an.github.ui.adapter.GithubListAdapter;
import com.an.github.ui.custom.recyclerview.RecyclerItemClickListener;
import com.an.github.ui.custom.recyclerview.RecyclerLayoutClickListener;
import com.an.github.ui.custom.recyclerview.RecyclerViewPaginator;
import com.an.github.ui.viewmodel.GithubListViewModel;
import com.an.github.utils.AnimUtils;
import com.an.github.utils.AppUtils;
import com.an.github.utils.NavigatorUtils;
import com.an.github.utils.ShareUtils;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

import javax.inject.Inject;

import dagger.android.AndroidInjection;

public class GithubListActivity extends AppCompatActivity implements RecyclerLayoutClickListener {

    @Inject
    ViewModelFactory viewModelFactory;

    private GithubListActivityBinding binding;
    private GithubListViewModel githubListViewModel;

    private GithubListAdapter githubListAdapter;
    private FilterListAdapter filterListAdapter;

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        AndroidInjection.inject(this);
        initialiseViewModel();
        initialiseView();
    }

    private void initialiseView() {
        binding = DataBindingUtil.setContentView(this, R.layout.activity_repo_list);
        setSupportActionBar(binding.mainToolbar.toolbar);

        filterListAdapter = new FilterListAdapter(Arrays.asList(getResources().getStringArray(R.array.list_filters)));
        binding.filterList.setLayoutManager(new LinearLayoutManager(getApplicationContext(), LinearLayoutManager.HORIZONTAL, false));
        binding.filterList.addOnItemTouchListener(new RecyclerItemClickListener(getApplicationContext(), (parentView, childView, position) -> {
            filterListAdapter.updateSelection(position);
            githubListAdapter.getFilter().filter(filterListAdapter.getItem(position));
        }));

        binding.recyclerView.setLayoutManager(new LinearLayoutManager(getApplicationContext()));
        githubListAdapter = new GithubListAdapter(getApplicationContext(), this);
        binding.recyclerView.setAdapter(githubListAdapter);
        binding.recyclerView.addOnScrollListener(new RecyclerViewPaginator(binding.recyclerView) {
            @Override
            public boolean isLastPage() {
                return githubListViewModel.isLastPage();
            }

            @Override
            public void loadMore() {
                githubListViewModel.fetchRepositories();
            }
        });

        /* This is to handle configuration changes:
         * during configuration change, when the activity
         * is recreated, we check if the viewModel
         * contains the list data. If so, there is no
         * need to call the api or load data from cache again */
        if(githubListViewModel.getRepositories().isEmpty()) {
            displayLoader();
            githubListViewModel.fetchRepositories();
        }
        else animateView(githubListViewModel.getRepositories());
    }


    private void initialiseViewModel() {
        githubListViewModel = ViewModelProviders.of(this, viewModelFactory).get(GithubListViewModel.class);
        githubListViewModel.getRepositoryListLiveData().observe(this, repositories -> {
            if(githubListAdapter.getItemCount() == 0) {
                if(!repositories.isEmpty()) {
                    animateView(repositories);

                } else displayEmptyView();

            } else if(!repositories.isEmpty()) displayDataView(repositories);
        });
    }

    private void displayLoader() {
        binding.viewLoader.rootView.setVisibility(View.VISIBLE);
    }

    private void hideLoader() {
        binding.viewLoader.rootView.setVisibility(View.GONE);
    }


    private void animateView(List<GithubEntity> repositories) {
        hideLoader();
        AnimUtils.slideView(binding.filterLayout, binding.filterList, filterListAdapter);
        displayDataView(repositories);
        binding.recyclerView.scheduleLayoutAnimation();
    }

    private void displayDataView(List<GithubEntity> repositories) {
        binding.viewEmpty.emptyContainer.setVisibility(View.GONE);
        githubListAdapter.setItems(repositories);
    }

    private void displayEmptyView() {
        hideLoader();
        binding.viewEmpty.emptyContainer.setVisibility(View.VISIBLE);
    }

    @Override
    public void redirectToDetailScreen(View imageView, View titleView, View revealView, View languageView, GithubEntity githubEntity) {
        NavigatorUtils.redirectToDetailScreen(this, githubEntity,
                ActivityOptionsCompat.makeSceneTransitionAnimation(this, AppUtils.getTransitionElements(
                        getApplicationContext(), imageView, titleView, revealView, languageView
                )));
    }

    @Override
    public void sharePost(GithubEntity githubEntity) {
        ShareUtils.shareUrl(this, githubEntity.getHtmlUrl());
    }
}

File: app/src/main/java/com/an/github/ui/activity/GithubListActivity.kt
package com.an.github.ui.activity

import android.os.Bundle
import android.view.View
import androidx.appcompat.app.AppCompatActivity
import com.an.github.data.local.entity.GithubEntity
import com.an.github.factory.ViewModelFactory
import com.an.github.ui.adapter.FilterListAdapter
import com.an.github.ui.adapter.GithubListAdapter
import com.an.github.ui.custom.recyclerview.RecyclerLayoutClickListener
import com.an.github.ui.viewmodel.GithubListViewModel
import javax.inject.Inject

class GithubListActivity : AppCompatActivity(), RecyclerLayoutClickListener {

    @Inject
    internal lateinit var viewModelFactory: ViewModelFactory

//    private lateinit var binding: GithubListActivityBinding
    private lateinit var githubListViewModel: GithubListViewModel

    private lateinit var githubListAdapter: GithubListAdapter
    private lateinit var filterListAdapter: FilterListAdapter

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
//        AndroidInjection.inject(this)
        initialiseViewModel()
        initialiseView()
    }

    private fun initialiseView() {
//        binding = DataBindingUtil.setContentView(this, R.layout.activity_repo_list)
//        setSupportActionBar(binding.mainToolbar.toolbar)
//
//        filterListAdapter = FilterListAdapter(Arrays.asList(*resources.getStringArray(R.array.list_filters)))
//        binding.filterList.layoutManager = LinearLayoutManager(applicationContext, LinearLayoutManager.HORIZONTAL, false)
//        binding.filterList.addOnItemTouchListener(RecyclerItemClickListener(applicationContext,
//                object : RecyclerItemClickListener.OnRecyclerViewItemClickListener {
//                    override fun onItemClick(parentView: View, childView: View, position: Int) {
//                        filterListAdapter.updateSelection(position)
//                        githubListAdapter.filter.filter(filterListAdapter.getItem(position))
//                    }
//
//                }))
//
//        binding.recyclerView.layoutManager = LinearLayoutManager(applicationContext)
//        githubListAdapter = GithubListAdapter(applicationContext, this)
//        binding.recyclerView.adapter = githubListAdapter
//        binding.recyclerView.addOnScrollListener(object : RecyclerViewPaginator(binding.recyclerView) {
//            override val isLastPage: Boolean
//                get() = githubListViewModel.isLastPage()
//
//            override fun loadMore() {
//                githubListViewModel.fetchRepositories()
//            }
//        })
//
//        /* This is to handle configuration changes:
//         * during configuration change, when the activity
//         * is recreated, we check if the viewModel
//         * contains the list data. If so, there is no
//         * need to call the api or load data from cache again */
//        if (githubListViewModel.getRepositories().isEmpty()) {
//            displayLoader()
//            githubListViewModel.fetchRepositories()
//        } else
//            animateView(githubListViewModel.getRepositories())
    }


    private fun initialiseViewModel() {
//        githubListViewModel = ViewModelProviders.of(this, viewModelFactory).get(GithubListViewModel::class.java)
//        githubListViewModel.getRepositoryLiveData().observe(this, Observer { repositories ->
//            if (githubListAdapter.itemCount == 0) {
//                repositories.withNotNullNorEmpty {
//                    animateView(repositories)
//                } .otherwise {
//                    displayEmptyView()
//                }
//
//            } else {
//                repositories.withNotNullNorEmpty {
//                    displayDataView(repositories)
//                }
//            }
//        })
    }

    private fun displayLoader() {
//        binding.viewLoader.rootView.visibility = View.VISIBLE
    }

    private fun hideLoader() {
//        binding.viewLoader.rootView.visibility = View.GONE
    }


    private fun animateView(repositories: List<GithubEntity>?) {
//        hideLoader()
//        AnimUtils.slideView(binding.filterLayout, binding.filterList, filterListAdapter)
//        displayDataView(repositories)
//        binding.recyclerView.scheduleLayoutAnimation()
    }

    private fun displayDataView(repositories: List<GithubEntity>?) {
//        binding.viewEmpty.emptyContainer.visibility = View.GONE
//        githubListAdapter.setItems(repositories)
    }

    private fun displayEmptyView() {
        hideLoader()
//        binding.viewEmpty.emptyContainer.visibility = View.VISIBLE
    }

    override fun redirectToDetailScreen(imageView: View, titleView: View, revealView: View, languageView: View, githubEntity: GithubEntity) {
//        NavigatorUtils.redirectToDetailScreen(this, githubEntity,
//                ActivityOptionsCompat.makeSceneTransitionAnimation(this, *AppUtils.getTransitionElements(
//                        applicationContext, imageView, titleView, revealView, languageView
//                )))
    }

    override fun sharePost(githubEntity: GithubEntity) {
//        ShareUtils.shareUrl(this, githubEntity.htmlUrl)
    }
}
